### PR TITLE
Create trip dashboard

### PR DESCRIPTION
**Pull request by yhelen**

### CODE REVIEW

File: app/models/student.rb [hunk 1]
```diff
@@ -56,6 +56,44 @@ def valid_forms(type)
...
+      stims.valid_for(year)
+    when :safety_tests
+      safety_tests.valid_for(year)
+    end
+  end
+
+  # A student will never have a permissions slip with
+  # the same type and trip as another permission slip
+  def find_permission_slip(type, trip_id)
+    permission_slips.where(type: type, trip_id: trip_id)
```
Comment (justinkim) [inline app/models/student.rb]: `where` will return a Relation, rather than a single `PermissionSlip`

File: app/models/student.rb [hunk 2]
```diff
@@ -56,6 +56,44 @@ def valid_forms(type)
...
+  end
+
+  # A student will never have two payments for the same
+  # payment requirement
+  def find_payment(required_payment_id)
+    payments.where(required_payment_id: required_payment_id)
```
Comment (justinkim) [inline app/models/student.rb]: `where` will return a Relation, rather than a single `Payment`

File: app/models/student.rb [hunk 3]
```diff
@@ -56,6 +56,44 @@ def valid_forms(type)
...
+  end
+
+  def eligible_for_trip(trip)
+    year = trip.end_date.year
```
Comment (justinkim) [inline app/models/student.rb]: This won't work in some cases like preseason events. Imagine an event in November 2017 like Brunswick. The students should have forms for the 2018 year.

File: app/models/student.rb [hunk 4]
```diff
@@ -56,6 +56,44 @@ def valid_forms(type)
...
+    has_consented_stim = valid_forms_for_year(:stims, year).consented.any?
+    has_signed_safety_test =
+      valid_forms_for_year(:safety_tests, year).signed.any?
+    puts valid_forms_for_year(:team_dues, year).any?
```
Comment (justinkim) [inline app/models/student.rb]: puts?

File: app/views/trips/_dashboard_button.html.haml [hunk 5]
```diff
@@ -0,0 +1,36 @@
+- if column_type == :payment
+  - id = "payment#{student.id}_#{required_payment.id}"
```
Comment (justinkim) [inline app/views/trips/_dashboard_button.html.haml]: Use [`dom_id`](https://apidock.com/rails/ActionView/RecordIdentifier/dom_id)

File: app/views/trips/_dashboard_button.html.haml [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+- else
+  - id = "#{column_type.to_s}#{student.id}"
```
Comment (justinkim) [inline app/views/trips/_dashboard_button.html.haml]: Use [`dom_id`](https://apidock.com/rails/ActionView/RecordIdentifier/dom_id)

File: app/views/payments/create.js.erb [hunk 7]
```diff
@@ -0,0 +1 @@
+$(<%= "payment#{@payment.student_id}_#{@payment.required_payment_id}" %>).replaceWith("<%=j render partial: 'trips/dashboard_button', locals: { column_type: "payment".to_sym, required_payment: @payment.required_payment, student: @payment.student } %>")
```
Comment (justinkim) [inline app/views/payments/create.js.erb]: I would try to make this span multiple lines so it's easier to read.

File: app/views/permission_slips/create.js.erb [hunk 8]
```diff
@@ -0,0 +1 @@
+$(<%= "#{@permission_slip.type.downcase}_permission#{@permission_slip.student_id}" %>).replaceWith("<%=j render partial: 'trips/dashboard_button', locals: { column_type: "#{@permission_slip.type.downcase}_permission".to_sym, student: @permission_slip.student, trip: @permission_slip.trip } %>")
```
Comment (justinkim) [inline app/views/permission_slips/create.js.erb]: I would try to make this span multiple lines so it's easier to read.

File: app/views/trips/_dashboard_button.html.haml [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+            student_id: student.id})
+
+  - else
+    - valid_button = false
+    - path = nil
+
+  - if valid_button
+    .btn.active.btn-success= "✔"
+  - elsif path != nil
+    = button_to "✘", path, method: :post, remote: true,
```
Comment (justinkim) [inline app/views/trips/_dashboard_button.html.haml]: What action is this path for? The button will have an x, but the method is post on the member.

File: test/models/medical_test.rb [hunk 10]
```diff
@@ -9,6 +9,24 @@ class MedicalTest < ActiveSupport::TestCase
     refute Medical.new(date: Date.today - 2.years).good?
   end
 
+  test "medical from valid from trip start to end date is good for trip" do
```
Comment (justinkim) [inline test/models/medical_test.rb]: I'm not sure I understand the test name. Maybe `a medical that is valid from a trip's start to end date is good for the trip`

File: app/views/trips/_dashboard_button.html.haml (hunk 9 above)
Comment (yhelen) [reply app/views/trips/_dashboard_button.html.haml]: I'm confused by this question, can you clarify? The path is a variable that is defined in the conditionals above.

Comment (justinkim) [reply app/views/trips/_dashboard_button.html.haml]: The path set above will route to the `create` action, but your button has an x on it, which makes it seem like it's to delete the model.

Comment (justinkim) [reply app/views/trips/_dashboard_button.html.haml]: nvm I'm dumb

**Pull request merged**